src/rslice.rs

```diff
@@ -0,0 +1,184 @@
+/// rustls_slice_bytes<'a> as an owned object. We can then expose slices of
+/// that Vec to C.
+pub(crate) struct VecSliceBytes<'a>(Vec<rustls_slice_bytes<'a>>);
+
+impl<'a> VecSliceBytes<'a> {
+    /// Build a VecSliceBytes that refers to `input` and can live as long as
+    /// it does.
+    fn new(input: &'a Vec<Vec<u8>>) -> Self {
+        let mut vv: Vec<rustls_slice_bytes> = vec![];
+        for v in input {
+            let v: &[u8] = v.as_ref();
+            vv.push(v.into());
+        }
+        VecSliceBytes(vv)
+    }
+}
+
+/// A read-only view of a slice of Rust byte slices.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API. The `data` is an array of `rustls_slice_bytes`
+/// structures with `len` elements.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not retain copies of any of the
+/// pointers beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_slice_bytes<'a> {
+    data: *const rustls_slice_bytes<'a>,
+    len: size_t,
+    phantom: PhantomData<&'a [&'a [u8]]>,
+}
+
+impl<'a> From<&'a VecSliceBytes<'a>> for rustls_slice_slice_bytes<'a> {
+    fn from(input: &'a VecSliceBytes<'a>) -> Self {
+        rustls_slice_slice_bytes {
+            data: input.0.as_ptr(),
+            len: input.0.len(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// A read-only view on a Rust `&str`. The contents are guaranteed to be valid
+/// UTF-8. As an additional guarantee on top of Rust's normal UTF-8 guarantee,
+/// a `rustls_str` is guaranteed not to contain internal NUL bytes, so it is
+/// safe to interpolate into a C string or compare using strncmp. Keep in mind
+/// that it is not NUL-terminated.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not retain copies of any of the
+/// pointers beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_str<'a> {
+    pub data: *const c_char,
+    pub len: size_t,
+    phantom: PhantomData<&'a str>,
+}
+
+impl<'a> TryFrom<&'a str> for rustls_str<'a> {
+    type Error = ();
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        if s.contains('\0') {
+            return Err(());
+        }
+        Ok(rustls_str {
+            data: s.as_ptr() as *const c_char,
+            len: s.len() as size_t,
+            phantom: PhantomData,
+        })
+    }
+}
+
+/// An owned Vec<rustls_str>, where the inner slices have lifetime 'a.
+/// See comment on VecSliceBytes for more information.
+pub(crate) struct VecStr<'a>(Vec<rustls_str<'a>>);
```

Comment:
It seems like `VecStr` and `VecSliceBytes` are manual monomorphizations of `Vec<T>`. But I think `VecStr` (for instance) only gets used in `rustls_slice_str::from<&'a VecStr<'a>>`, so it's never exposed to C. What I wonder is: could you implement `From<&'a Vec<rustls_str<'a>>` on `rustls_slice_str`, move the implementation of `VecStr::new` into `rustls_slice_str::from` and eliminate the need for `VecStr` altogether? I think that's pretty much what you did for `rustls_slice_u16`.